### PR TITLE
Issue201/remote chaincode build

### PR DESCRIPTION
**Pull request by prjayach**
#201 Support for allowing remote URL for go chaincode.

Fix for ledger lock issue in chaincode tests.
Rebased with latest master. All tests pass except the peer tests which are still failing in master.

### CODE REVIEW

**Comment by prjayach** [comment]
Signed-off by: Praveen Jayachandran prjayach@in.ibm.com

**Comment by muralisrini** [comment]
Looks good for most part Praveen. Time has come though to promote "getChaincodeID" ? I think @srderson also mentioned a need for such a function elsewhere. If we move it into an util or something, we can reuse that code. Noticed the code is duped in a few places already.

**Comment by prjayach** [comment]
@muralisrini Yes, I considered this. I have consolidated multiple places where we were constructing VM name into one function GetVMName in container/controller.go. The chaincodeID is a substring of this name and I think its best to keep a separate function for this. I didn't want to import package vm within package chaincode, so left this method within chaincode. If there is a better suggestion of where we can place these methods, I am all for it.

**Comment by srderson** [comment]
There's a util package with utils.go. That could be a possible location.

File: core/container/externalbuilder/copy.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+)
+
+// MoveOrCopyDir attempts to copy src to dest by firstly trying to move, then copy upon failure.
+func MoveOrCopyDir(logger *flogging.FabricLogger, srcroot, destroot string) error {
+	mvErr := os.Rename(srcroot, destroot)
+	if mvErr == nil {
+		return nil
+	}
+
+	logger.Warnf("Failed to move %s to %s: %s, try copy instead", srcroot, destroot, mvErr)
```
Comment (jyellick) [inline core/container/externalbuilder/copy.go]: Warning seems a bit severe here, no?  I would expect that in most production environment the TMPDIR will not be on persistent storage while the chaincode persistence dir necessarily would be.  Maybe turn this down to Debug?

File: core/container/externalbuilder/copy.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+	info, err := os.Stat(srcroot)
+	if err != nil {
+		return errors.WithMessagef(err, "failed to stat")
```
Comment (jyellick) [inline core/container/externalbuilder/copy.go]: Very much a nit, but no parameters and a `WithMessagef`.  Wouldn't have even commented, but at line 33 below you choose a `WithMessage`

**Pull request merged**